### DISCUSSION TITLE
tasks.json vs launch.json, and test

### DISCUSSION THREAD

**Discussion by hyangah** [microsoft/vscode-discussions#1206] [Extension Development QnA] [answered]
Hi, I am the Go extension maintainer.

Our extension provides test functionalities by running the `go test` command (using the node.js child process) and presenting the results using the Test API.

As users want more flexibilities and controls over how they run the command, I am investigating whether we can use VS Code Task API in running `go test` instead of working with node.js child process directly. Then, I hope users can offer their customization by defining a task in test group using their tasks.json, too.

On the other hand, there is also `launch.json`. It is primarily for debugging and can carry debug-specific settings. But it also offers the `noDebug` option.

Users often use `run test` and `debug test` features together. They expect the identical setup (except the debug-specific part) to be used when switching between them. Repeating the similar setting in `tasks.json` and `launch.json` doesn't seem ideal. 

How do other language&frame extensions handle test&debug configurations? Do they use `tasks.json` or `launch.json` for test configuration? Or do they define extension configuration instead?

**Comment by isidorn** [comment]
Hi Go friends 👋 

@connor4312 might provide guidance here

**Comment by connor4312** [comment, marked as answer]
There's been quite some discussion on this in Python in https://github.com/microsoft/vscode-python/issues/21845.

The approach I took with testing in VS Code initially was to avoid adding additional config files, e.g. no debug-esque launch.json. The reasoning for this was that configuring launch.json is constantly a point for users, and many test runners already have their own configuration files available to configure them.

For now the general approach is to use workspace settings: that's what Python does and also what I do in [my own test extension](https://github.com/connor4312/nodejs-testing) for example. We may add some extra features here in the future, such as making tests configurable in the launch.json. Note that you could also source data from the launch.json (calling `vscode.workspace.getConfiguration('launch')` will get you the contents of launch.json) if you want to read debug settings from there.

I don't anticipate changing how the test child process is owned at any point, because extensions have widely varying lifecycle needs.
